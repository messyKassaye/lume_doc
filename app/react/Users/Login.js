import PropTypes from 'prop-types';
import React from 'react';
import { browserHistory } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Field, LocalForm, actions as formActions } from 'react-redux-form';

import { Icon } from 'UI';

import { t } from 'app/I18N';
import { isClient } from 'app/utils';
import { reconnectSocket } from 'app/socket';
import RouteHandler from 'app/App/RouteHandler';
import { reloadThesauri } from 'app/Thesauri/actions/thesaurisActions';

import auth from 'app/Auth';

const reloadHome = () => {
  window.location.assign('/');
};

export class Login extends RouteHandler {
  constructor(props, context) {
    super(props, context);
    this.state = { error: false, error2fa: false, recoverPassword: false, tokenRequired: false };
    this.submit = this.submit.bind(this);
    this.setLogin = this.setLogin.bind(this);
    this.setRecoverPassword = this.setRecoverPassword.bind(this);
  }

  attachDispatch(dispatch) {
    this.formDispatch = dispatch;
  }

  submit(credentials) {
    if (this.state.recoverPassword) {
      return this.recoverPassword(credentials.username);
    }

    return this.login(credentials);
  }

  recoverPassword(email) {
    this.setState({ recoverPassword: false, error: false });
    this.formDispatch(formActions.reset('loginForm'));
    return this.props.recoverPassword(email);
  }

  resolveSuccessfulLogin() {
    if (this.props.private) {
      reloadHome();
      return;
    }
    reconnectSocket();
    this.props.reloadThesauris();
    browserHistory.push('/');
  }

  async login(credentials) {
    try {
      await this.props.login(credentials);
      this.resolveSuccessfulLogin();
    } catch (err) {
      if (!this.state.tokenRequired && err.status === 409) {
        this.setState({ tokenRequired: true });
      } else {
        const { tokenRequired } = this.state;
        this.formDispatch(formActions.change('loginForm.token', undefined));
        const error2fa = tokenRequired;
        this.setState({ error: true, tokenRequired, error2fa });
      }
    }
  }

  setRecoverPassword() {
    this.formDispatch(formActions.reset('loginForm'));
    this.setState({ recoverPassword: true, error: false });
  }

  setLogin() {
    this.formDispatch(formActions.reset('loginForm'));
    this.setState({ recoverPassword: false, tokenRequired: false, error: false, error2fa: false });
  }

  render() {
    let submitLabel = this.state.recoverPassword
      ? t('System', 'Send recovery email')
      : t('System', 'Login button', 'Login');

    if (this.state.tokenRequired) {
      submitLabel = t('System', 'Verify');
    }

    return (
      <div className="content login-content">
        <div className="row">
          <div className="col-xs-12 col-sm-4 col-sm-offset-4" suppressHydrationWarning>
            <h1 className="login-title">
              <img src="/public/logo.svg" title="uwazi" alt="uwazi" />
            </h1>
            {isClient && (
              <LocalForm
                onSubmit={this.submit}
                model="loginForm"
                getDispatch={dispatch => {
                  this.formDispatch = dispatch;
                }}
              >
                {!this.state.tokenRequired && (
                  <>
                    <div
                      className={`form-group login-email${this.state.error ? ' has-error' : ''}`}
                    >
                      <Field model=".username">
                        <label className="form-group-label" htmlFor="username">
                          {this.state.recoverPassword ? t('System', 'Email') : t('System', 'User')}
                        </label>
                        <input type="text" name="username" id="username" className="form-control" />
                      </Field>
                    </div>
                    <div
                      className={`form-group login-password ${this.state.error ? 'has-error' : ''}${
                        this.state.recoverPassword ? ' is-hidden' : ''
                      }`}
                    >
                      <label className="form-group-label" htmlFor="password">
                        {t('System', 'Password')}
                      </label>
                      <Field model=".password">
                        <input
                          type="password"
                          name="password"
                          id="password"
                          className="form-control"
                        />
                      </Field>
                      <div className="form-text">
                        {this.state.error && <span>{t('System', 'Login failed')} - </span>}
                        <span
                          title={t('System', 'Forgot Password?', null, false)}
                          onClick={this.setRecoverPassword}
                          className={`button forgot-password ${
                            this.state.error ? 'label-danger' : ''
                          }`}
                        >
                          {t('System', 'Forgot Password?')}
                        </span>
                      </div>
                    </div>
                  </>
                )}
                {this.state.tokenRequired && (
                  <div
                    className={`form-group login-token${this.state.error2fa ? ' has-error' : ''}`}
                  >
                    <h5>{t('System', 'Two-step verification')}</h5>
                    <Field model=".token">
                      <label className="form-group-label" htmlFor="token">
                        {t('System', 'Authentication code')}
                      </label>
                      <input type="text" name="token" id="token" className="form-control" />
                      <div className="form-text">
                        <p>
                          {this.state.error2fa && (
                            <>
                              <Icon icon="exclamation-triangle" />
                              <span>{t('System', 'Two-factor verification failed')}</span>
                            </>
                          )}
                        </p>
                        <p>
                          Open the two-factor Authenticator app on your device <br />
                          to view your authentication code and verify your identity.
                        </p>
                        <p>
                          <span
                            onClick={this.setLogin}
                            className={`button forgot-password ${
                              this.state.error2fa ? 'label-danger' : ''
                            }`}
                          >
                            {t('System', 'Return to login')}
                          </span>
                        </p>
                      </div>
                    </Field>
                  </div>
                )}
                <p>
                  <button
                    type="submit"
                    className={`btn btn-block btn-lg ${
                      this.state.recoverPassword ? 'btn-success' : 'btn-primary'
                    }`}
                  >
                    {submitLabel}
                  </button>
                </p>
                {this.state.recoverPassword && (
                  <div className="form-text">
                    <span
                      title={t('System', 'Cancel', null, false)}
                      onClick={this.setLogin}
                      className="button cancel"
                    >
                      {t('System', 'Cancel')}
                    </span>
                  </div>
                )}
              </LocalForm>
            )}
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func,
  recoverPassword: PropTypes.func,
  reloadThesauris: PropTypes.func,
};

export function mapStateToProps({ settings }) {
  return {
    private: settings.collection.get('private'),
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      login: auth.actions.login,
      recoverPassword: auth.actions.recoverPassword,
      reloadThesauris: reloadThesauri,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
