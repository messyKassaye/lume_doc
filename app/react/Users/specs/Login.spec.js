/**
 * @jest-environment jsdom
 */

import React from 'react';
import { shallow } from 'enzyme';
import { browserHistory } from 'react-router';
import { actions as formActions } from 'react-redux-form';

import { Login } from '../Login.js';

describe('Login', () => {
  let component;
  let instance;
  let props;
  let formDispatch;

  const render = () => {
    const context = { store: { getState: () => ({}) } };
    component = shallow(<Login {...props} />, { context });
    instance = component.instance();
    instance.attachDispatch(formDispatch);
  };

  const expectState = (state, expected) => {
    expect(state).toEqual(expect.objectContaining(expected));
  };

  beforeEach(() => {
    formDispatch = jasmine.createSpy('formDispatch');
    props = {
      login: jasmine.createSpy('login').and.returnValue(Promise.resolve()),
      recoverPassword: jasmine.createSpy('recoverPassword').and.returnValue(Promise.resolve()),
      notify: jasmine.createSpy('notify'),
      reloadThesauris: jasmine.createSpy('reloadThesauris'),
      routes: [],
    };
    spyOn(formActions, 'reset').and.callFake(formName => formName);
    render();
  });

  describe('on instance', () => {
    it('should set state', () => {
      expect(instance.state).toEqual({
        error: false,
        error2fa: false,
        recoverPassword: false,
        tokenRequired: false,
      });
    });

    it('should render the component with the login form', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('submit()', () => {
    it('should send the credentials', done => {
      instance
        .submit('credentials')
        .then(() => {
          expect(props.login).toHaveBeenCalledWith('credentials');
          done();
        })
        .catch(done.fail);
    });

    describe('when recoverPassword is true', () => {
      it('should call recoverPassword with the email', async () => {
        instance.state.recoverPassword = true;
        await instance.submit({ username: 'email@recover.com' });
        expect(props.recoverPassword).toHaveBeenCalledWith('email@recover.com');
        expect(formDispatch).toHaveBeenCalledWith('loginForm');
        expectState(instance.state, {
          error: false,
          error2fa: false,
          recoverPassword: false,
          tokenRequired: false,
        });
      });
    });

    describe('on response success', () => {
      it('should reload thesauris and go to home', async () => {
        spyOn(browserHistory, 'push');
        expect(props.reloadThesauris).not.toHaveBeenCalled();
        await instance.submit('credentials');

        expect(props.reloadThesauris).toHaveBeenCalled();
        expect(browserHistory.push).toHaveBeenCalledWith('/');
      });

      describe('when the instance is private', () => {
        it('should call reloadHome', done => {
          props.private = true;
          render();

          spyOn(window.location, 'assign');

          instance
            .submit('credentials')
            .then(() => {
              expect(window.location.assign).toHaveBeenCalledWith('/');
              done();
            })
            .catch(done.fail);
        });
      });
    });

    describe('on response failure', () => {
      const prepareLoginResponse = response => {
        props.login = jasmine.createSpy('login').and.returnValue(response);
        render();
      };

      const response409 = () => {
        const error = new Error('Conflict');
        error.status = 409;
        return error;
      };

      describe('when authorization conflict (2fa required)', () => {
        it('should not set error and flag "tokenRequired"', async () => {
          prepareLoginResponse(Promise.reject(response409()));
          await instance.submit(new Event('submit'));
          expectState(instance.state, { error: false, tokenRequired: true });
          expect(component).toMatchSnapshot();
        });
      });

      describe('when authorization failure', () => {
        beforeEach(() => {
          prepareLoginResponse(Promise.reject(new Error({ status: 401 })));
        });

        it('should set error upon login failure', async () => {
          await instance.submit(new Event('submit'));
          expectState(instance.state, { error: true, tokenRequired: false, error2fa: false });
        });

        it('should set error2fa on token failure, and reset loginForm.token value', async () => {
          instance.state.tokenRequired = true;
          await instance.submit(new Event('submit'));
          expectState(instance.state, { error: true, tokenRequired: true, error2fa: true });
          expect(instance.formDispatch).toHaveBeenCalledWith(
            expect.objectContaining({ model: 'loginForm.token', value: undefined })
          );
          expect(component).toMatchSnapshot();
        });
      });
    });
  });

  describe('setRecoverPassword()', () => {
    it('should ser recoverPassword true, and error false', () => {
      instance.setRecoverPassword();
      expect(instance.state.error).toBe(false);
      expect(instance.state.recoverPassword).toBe(true);
    });
  });
});
