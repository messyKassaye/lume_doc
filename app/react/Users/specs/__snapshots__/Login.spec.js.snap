// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login on instance should render the component with the login form 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
      suppressHydrationWarning={true}
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-email"
        >
          <Modeled(Connect(Field))
            model=".username"
          >
            <label
              className="form-group-label"
              htmlFor="username"
            >
              User
            </label>
            <input
              className="form-control"
              id="username"
              name="username"
              type="text"
            />
          </Modeled(Connect(Field))>
        </div>
        <div
          className="form-group login-password "
        >
          <label
            className="form-group-label"
            htmlFor="password"
          >
            Password
          </label>
          <Modeled(Connect(Field))
            model=".password"
          >
            <input
              className="form-control"
              id="password"
              name="password"
              type="password"
            />
          </Modeled(Connect(Field))>
          <div
            className="form-text"
          >
            <span
              className="button forgot-password "
              onClick={[Function]}
              title="Forgot Password?"
            >
              Forgot Password?
            </span>
          </div>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            Login
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;

exports[`Login submit() on response failure when authorization conflict (2fa required) should not set error and flag "tokenRequired" 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
      suppressHydrationWarning={true}
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-token"
        >
          <h5>
            Two-step verification
          </h5>
          <Modeled(Connect(Field))
            model=".token"
          >
            <label
              className="form-group-label"
              htmlFor="token"
            >
              Authentication code
            </label>
            <input
              className="form-control"
              id="token"
              name="token"
              type="text"
            />
            <div
              className="form-text"
            >
              <p />
              <p>
                Open the two-factor Authenticator app on your device 
                <br />
                to view your authentication code and verify your identity.
              </p>
              <p>
                <span
                  className="button forgot-password "
                  onClick={[Function]}
                >
                  Return to login
                </span>
              </p>
            </div>
          </Modeled(Connect(Field))>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            Verify
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;

exports[`Login submit() on response failure when authorization failure should set error2fa on token failure, and reset loginForm.token value 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
      suppressHydrationWarning={true}
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-token has-error"
        >
          <h5>
            Two-step verification
          </h5>
          <Modeled(Connect(Field))
            model=".token"
          >
            <label
              className="form-group-label"
              htmlFor="token"
            >
              Authentication code
            </label>
            <input
              className="form-control"
              id="token"
              name="token"
              type="text"
            />
            <div
              className="form-text"
            >
              <p>
                <Connect(Icon)
                  icon="exclamation-triangle"
                />
                <span>
                  Two-factor verification failed
                </span>
              </p>
              <p>
                Open the two-factor Authenticator app on your device 
                <br />
                to view your authentication code and verify your identity.
              </p>
              <p>
                <span
                  className="button forgot-password label-danger"
                  onClick={[Function]}
                >
                  Return to login
                </span>
              </p>
            </div>
          </Modeled(Connect(Field))>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            Verify
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;
